azurerm/data_source_resources.go

```diff
@@ -0,0 +1,177 @@
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/tags"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/timeouts"
+)
+
+func dataSourceArmResources() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceArmResourcesRead,
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			"resource_group_name": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			"type": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+
+			"required_tags": tags.Schema(),
+
+			"resources": {
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"name": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"id": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"type": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"location": azure.SchemaLocationForDataSource(),
+						"tags":     tags.SchemaDataSource(),
+					},
+				},
+			},
+		},
+	}
+}
+
+func dataSourceArmResourcesRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).Resource.ResourcesClient
+	ctx, cancel := timeouts.ForRead(meta.(*ArmClient).StopContext, d)
+	defer cancel()
+
+	resourceGroupName := d.Get("resource_group_name").(string)
+	resourceName := d.Get("name").(string)
+	resourceType := d.Get("type").(string)
+	requiredTags := d.Get("required_tags").(map[string]interface{})
+
+	if resourceGroupName == "" && resourceName == "" && resourceType == "" {
+		return fmt.Errorf("At least one of `name`, `resource_group_name` or `type` must be specified")
+	}
+
+	var filter string
+
+	if resourceGroupName != "" {
+		v := fmt.Sprintf("resourceGroup eq '%s'", resourceGroupName)
+		filter = filter + v
+	}
+
+	if resourceName != "" {
+		if strings.Contains(filter, "eq") {
+			filter = filter + " and "
+		}
+		v := fmt.Sprintf("name eq '%s'", resourceName)
+		filter = filter + v
+	}
+
+	if resourceType != "" {
+		if strings.Contains(filter, "eq") {
+			filter = filter + " and "
+		}
+		v := fmt.Sprintf("resourceType eq '%s'", resourceType)
+		filter = filter + v
+	}
+
+	resources := make([]map[string]interface{}, 0)
+	resource, err := client.ListComplete(ctx, filter, "", nil)
+	if err != nil {
+		return fmt.Errorf("Error getting resources: %+v", err)
+	}
+
+	for resource.NotDone() {
+		res := resource.Value()
+		if res.ID == nil {
+			continue
+		}
+
+		// currently its not supported to use tags filter with other filters
+		// therefore we need to filter the resources manually.
+		tagMatches := 0
+		if res.Tags != nil {
+			for requiredTagName, requiredTagVal := range requiredTags {
+				for tagName, tagVal := range res.Tags {
+					if requiredTagName == tagName && requiredTagVal == *tagVal {
```

Comment:
since `tagVal` is a `*string` here this wants a nil-check:

```suggestion
					if requiredTagName == tagName && tagVal != nil && requiredTagVal == *tagVal {
```